### PR TITLE
Refactor metal backend for memory efficiency

### PR DESCRIPTION
**Pull request by Maroc-OS**
I hope this one is more clear than the previews PR.

changes made are minimal

### CODE REVIEW

**Comment by geohot** [comment]
There's still mixed whitespace/typing and functional changes.

File: tinygrad/runtime/ops_metal.py [hunk 1]
```diff
@@ -23,10 +22,14 @@ def synchronize(self):
 class RawMetalBuffer(RawBufferMapped):
   def __init__(self, size:int, dtype:DType):
     assert dtype != dtypes.float64, "metal doesn't support float64"
-    super().__init__(size, dtype, METAL.device.newBufferWithLength_options_(size*dtype.itemsize, Metal.MTLResourceStorageModeShared))
+    super().__init__(size, dtype, METAL.device.newBufferWithBytesNoCopy_length_options_deallocator_(None, size*dtype.itemsize, Metal.MTLResourceStorageModeShared, None))
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:25]: What if the backing object is freed?

File: tinygrad/runtime/ops_metal.py [hunk 2]
```diff
@@ -23,10 +22,14 @@ def synchronize(self):
...
   def __del__(self):
-    self._buf.release()
+    self.release()
     super().__del__()
+  def release(self):
```
Comment (geohot) [inline tinygrad/runtime/ops_metal.py:29]: Why does this have to be a function?

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/runtime/ops_metal.py             10               8
------------------------------------------------------------
lines added in the tinygrad folder: 2
```

File: tinygrad/runtime/ops_metal.py (hunk 1 above)
Comment (geohot) [reply tinygrad/runtime/ops_metal.py:25]: Wait this makes no sense. The pointer is None. Isn't this just the same thing?

**Comment by geohot** [comment]
I'm going to close this. I still don't understand what it gains, there's no Metal CI tests yet, and I don't like passing None to the no copy, it just seems like the wrong constructor.

**Pull request closed without merging**